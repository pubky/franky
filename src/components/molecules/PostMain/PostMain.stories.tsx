import type { Meta, StoryObj } from '@storybook/nextjs-vite';
import { PostMain } from './PostMain';

const meta: Meta<typeof PostMain> = {
  title: 'Molecules/Post/PostMain',
  component: PostMain,
  parameters: {
    layout: 'centered',
    backgrounds: {
      default: 'dark',
      values: [
        { name: 'dark', value: '#05050A' },
        { name: 'light', value: '#F2F2F7' },
      ],
    },
  },
  args: {
    avatarSrc: '/images/avatar.png',
    displayName: 'Satoshi Nakamoto',
    label: '1YXP...7R32',
    timeLabel: '15m',
    text: 'I have said it before: If you donâ€™t believe me or donâ€™t get it, I donâ€™t have time to try to convince you.',
    tags: [
      { label: 'bitcoin', count: 21, color: '#FF9900' },
      { label: 'based', count: 7, color: '#FF0000' },
      { label: 'ðŸ‘€', count: 5, color: '#00F0FF' },
    ],
    showAddButton: true,
    showEmojiPicker: true,
  },
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  render: (args) => (
    <div className="w-[720px]">
      <PostMain {...args} />
    </div>
  ),
};

export const Mobile: Story = {
  render: (args) => (
    <div className="w-[375px]">
      <PostMain {...args} />
    </div>
  ),
  parameters: {
    viewport: { defaultViewport: 'mobile1' },
    docs: {
      description: {
        story: 'Mobile layout with stacked Tags and Actions and tighter paddings.',
      },
    },
  },
};
