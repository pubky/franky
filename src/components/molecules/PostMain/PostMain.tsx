import * as React from 'react';
import * as Molecules from '@/molecules';
import * as Atoms from '@/atoms';
import * as Libs from '@/libs';

import type { PostTagsListTag } from '@/molecules/PostTagsList/PostTagsList';

export interface PostMainProps {
  className?: string;
  // Header
  avatarSrc?: string;
  avatarAlt?: string;
  displayName: string;
  label?: string;
  timeLabel?: string;
  // Content
  text: string;
  // Tags
  tags?: PostTagsListTag[];
  showAddButton?: boolean;
  showEmojiPicker?: boolean;
  showTagClose?: boolean;
  onTagClick?: (tag: PostTagsListTag, index: number) => void;
  onTagClose?: (tag: PostTagsListTag, index: number) => void;
  onTagAdd?: (label: string) => void;
  onEmojiClick?: () => void;
  // Actions
  tagCount?: number;
  replyCount?: number;
  repostCount?: number;
  isBookmarked?: boolean;
  onTagActionClick?: () => void;
  onReplyClick?: () => void;
  onRepostClick?: () => void;
  onBookmarkClick?: () => void;
  onMoreClick?: () => void;
}

export function PostMain({
  className,
  avatarSrc,
  avatarAlt,
  displayName,
  label,
  timeLabel,
  text,
  tags = [],
  showAddButton = true,
  showEmojiPicker = true,
  showTagClose = false,
  onTagClick,
  onTagClose,
  onTagAdd,
  onEmojiClick,
  tagCount = 5,
  replyCount = 7,
  repostCount = 3,
  isBookmarked = false,
  onTagActionClick,
  onReplyClick,
  onRepostClick,
  onBookmarkClick,
  onMoreClick,
}: PostMainProps) {
  return (
    <Atoms.Card className={Libs.cn('w-full', className)}>
      <Atoms.CardContent className="px-4 md:px-6 md:py-0">
        <div className="flex flex-col gap-4 md:gap-4 w-full">
          <Molecules.PostHeader
            avatarSrc={avatarSrc}
            avatarAlt={avatarAlt}
            displayName={displayName}
            label={label}
            timeLabel={timeLabel}
          />

          <Molecules.PostContent text={text} />

          <div className="flex flex-col md:flex-row items-stretch md:items-center gap-4  flex-wrap md:flex-nowrap justify-start md:justify-start">
            <Molecules.PostTagsList
              tags={tags}
              showInput={false}
              showAddButton={showAddButton}
              addMode
              showEmojiPicker={showEmojiPicker}
              showTagClose={showTagClose}
              onTagClick={onTagClick}
              onTagClose={onTagClose}
              onTagAdd={onTagAdd}
              onEmojiClick={onEmojiClick}
            />

            <Molecules.PostActionsBar
              tagCount={tagCount}
              replyCount={replyCount}
              repostCount={repostCount}
              isBookmarked={isBookmarked}
              onTagClick={onTagActionClick}
              onReplyClick={onReplyClick}
              onRepostClick={onRepostClick}
              onBookmarkClick={onBookmarkClick}
              onMoreClick={onMoreClick}
              className="justify-start w-full flex-1 md:justify-end"
            />
          </div>
        </div>
      </Atoms.CardContent>
    </Atoms.Card>
  );
}
