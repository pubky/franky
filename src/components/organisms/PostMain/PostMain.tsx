'use client';

import { useLiveQuery } from 'dexie-react-hooks';
import * as Core from '@/core';
import * as Libs from '@/libs';
import * as Atoms from '@/atoms';
import * as Molecules from '@/molecules';
import * as Organisms from '@/organisms';

export interface PostMainProps {
  postId: string;
  onClick?: () => void;
  className?: string;
}

export function PostMain({ postId, onClick, className }: PostMainProps) {
  // Fetch post tags
  const postTags = useLiveQuery(async () => {
    return await Core.PostTagsModel.findById(postId);
  }, [postId]);

  const tags =
    postTags?.tags.map((tag, index) => ({
      id: `${postId}-tag-${index}`,
      label: tag.label,
    })) || [];

  return (
    <div onClick={onClick} className="cursor-pointer">
      <Atoms.Card className={Libs.cn('w-full', className)}>
        <Atoms.CardContent className="px-4 md:px-6 md:py-0">
          <div className="flex flex-col gap-4 md:gap-4 w-full">
            <Organisms.PostHeader postId={postId} />

            <Organisms.PostContent postId={postId} />

            <div className="flex flex-col md:flex-row items-stretch md:items-center gap-4 flex-wrap md:flex-nowrap justify-start md:justify-start">
              <Molecules.PostTagsList
                tags={tags}
                showInput={false}
                showAddButton={false}
                addMode
                showEmojiPicker={false}
                showTagClose={false}
              />

              <Organisms.PostActionsBar postId={postId} className="justify-start w-full flex-1 md:justify-end" />
            </div>
          </div>
        </Atoms.CardContent>
      </Atoms.Card>
    </div>
  );
}
