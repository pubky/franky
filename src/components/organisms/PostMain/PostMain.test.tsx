import { render, screen, fireEvent } from '@testing-library/react';
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { PostMain } from './PostMain';
import { useLiveQuery } from 'dexie-react-hooks';

vi.mock('dexie-react-hooks', () => ({
  useLiveQuery: vi.fn(),
}));

// Use real libs, only stub cn for deterministic class joining
vi.mock('@/libs', async (importOriginal) => {
  const actual = await importOriginal<typeof import('@/libs')>();
  return {
    ...actual,
    cn: (...classes: (string | undefined)[]) => classes.filter(Boolean).join(' '),
  };
});

// Minimal atoms used by PostMain
vi.mock('@/atoms', () => ({
  Card: ({ children, className }: { children: React.ReactNode; className?: string }) => (
    <div data-testid="card" data-class-name={className}>
      {children}
    </div>
  ),
  CardContent: ({ children, className }: { children: React.ReactNode; className?: string }) => (
    <div data-testid="card-content" data-class-name={className}>
      {children}
    </div>
  ),
}));

// Stub organisms composed inside PostMain
vi.mock('@/organisms', () => ({
  PostHeader: ({ postId }: { postId: string }) => <div data-testid="post-header">PostHeader {postId}</div>,
  PostContent: ({ postId }: { postId: string }) => <div data-testid="post-content">PostContent {postId}</div>,
  PostActionsBar: ({ postId, className }: { postId: string; className?: string }) => (
    <div data-testid="post-actions" data-class-name={className}>
      Actions {postId}
    </div>
  ),
}));

// Stub molecules used by PostMain
vi.mock('@/molecules', () => ({
  PostTagsList: ({ tags }: { tags: Array<{ id: string; label: string }> }) => (
    <ul data-testid="post-tags-list">
      {tags.map((t) => (
        <li key={t.id}>{t.label}</li>
      ))}
    </ul>
  ),
}));

// Keep Core shape only for potential spying if needed
vi.mock('@/core', () => ({
  PostTagsModel: {
    findById: vi.fn(),
  },
}));

const mockUseLiveQuery = vi.mocked(useLiveQuery);

describe('PostMain', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders header, content, tags and actions', () => {
    mockUseLiveQuery.mockReturnValue({ tags: [{ label: 'alpha' }, { label: 'beta' }] });

    render(<PostMain postId="post-123" />);

    expect(screen.getByTestId('post-header')).toHaveTextContent('PostHeader post-123');
    expect(screen.getByTestId('post-content')).toHaveTextContent('PostContent post-123');
    expect(screen.getByTestId('post-tags-list')).toBeInTheDocument();
    expect(screen.getByText('alpha')).toBeInTheDocument();
    expect(screen.getByText('beta')).toBeInTheDocument();
    expect(screen.getByTestId('post-actions')).toBeInTheDocument();
  });

  it('invokes onClick handler when root is clicked', () => {
    mockUseLiveQuery.mockReturnValue({ tags: [] });
    const onClick = vi.fn();

    const { container } = render(<PostMain postId="post-abc" onClick={onClick} />);

    fireEvent.click(container.firstChild as HTMLElement);
    expect(onClick).toHaveBeenCalledTimes(1);
  });
});

describe('PostMain - Snapshots', () => {
  it('matches snapshot with tags', () => {
    mockUseLiveQuery.mockReturnValue({ tags: [{ label: 'alpha' }, { label: 'beta' }, { label: 'gamma' }] });

    const { container } = render(<PostMain postId="post-123" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot without tags', () => {
    mockUseLiveQuery.mockReturnValue({ tags: [] });

    const { container } = render(<PostMain postId="post-789" />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
