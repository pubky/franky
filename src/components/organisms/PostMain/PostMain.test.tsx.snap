// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PostMain - Snapshots > matches snapshot with tags 1`] = `
<div
  class="cursor-pointer"
>
  <div
    data-class-name="w-full"
    data-testid="card"
  >
    <div
      data-class-name="px-4 md:px-6 md:py-0"
      data-testid="card-content"
    >
      <div
        class="flex flex-col gap-4 md:gap-4 w-full"
      >
        <div
          data-testid="post-header"
        >
          PostHeader 
          post-123
        </div>
        <div
          data-testid="post-content"
        >
          PostContent 
          post-123
        </div>
        <div
          class="flex flex-col md:flex-row items-stretch md:items-center gap-4 flex-wrap md:flex-nowrap justify-start md:justify-start"
        >
          <ul
            data-testid="post-tags-list"
          >
            <li>
              alpha
            </li>
            <li>
              beta
            </li>
            <li>
              gamma
            </li>
          </ul>
          <div
            data-class-name="justify-start w-full flex-1 md:justify-end"
            data-testid="post-actions"
          >
            Actions 
            post-123
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PostMain - Snapshots > matches snapshot without tags 1`] = `
<div
  class="cursor-pointer"
>
  <div
    data-class-name="w-full"
    data-testid="card"
  >
    <div
      data-class-name="px-4 md:px-6 md:py-0"
      data-testid="card-content"
    >
      <div
        class="flex flex-col gap-4 md:gap-4 w-full"
      >
        <div
          data-testid="post-header"
        >
          PostHeader 
          post-789
        </div>
        <div
          data-testid="post-content"
        >
          PostContent 
          post-789
        </div>
        <div
          class="flex flex-col md:flex-row items-stretch md:items-center gap-4 flex-wrap md:flex-nowrap justify-start md:justify-start"
        >
          <ul
            data-testid="post-tags-list"
          />
          <div
            data-class-name="justify-start w-full flex-1 md:justify-end"
            data-testid="post-actions"
          >
            Actions 
            post-789
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
